File: src/c04_professional_classes/dependency_inversion/DependencyInversionRefactored.java
package c04_professional_classes.dependency_inversion;

import java.util.ArrayList;
import java.util.List;

/*
In this example:

- We create an IWorker interface that Worker and Robot implement.
- The Manager class depends on the IWorker abstraction, not specific worker implementations.
- This adheres to DIP, allowing the Manager class to work with any class that implements IWorker, making the system more flexible and easier to extend.
*/

interface IWorker {
    void work();
}

class Worker implements IWorker {
    @Override
    public void work() {
        System.out.println("Worker is working...");
    }
}

class Robot implements IWorker {
    @Override
    public void work() {
        System.out.println("Robot is working...");
    }
}

class Manager {
    private List<IWorker> workers = new ArrayList<>();

    public void addWorker(IWorker worker) {
        workers.add(worker);
    }

    public void manage() {
        for (IWorker worker : workers) {
            worker.work();
        }
    }
}

public class DependencyInversionRefactored {
    public static void main(String[] args) {
        Manager manager = new Manager();
        manager.addWorker(new Worker());
        manager.addWorker(new Robot());
        manager.manage();
    }
}

File: src/c04_professional_classes/matlab/DependencyInversionRefactored.m
function DependencyInversionRefactored()
    % In this example:
    % - We create an IWorker interface (abstract class) that Worker and Robot implement.
    % - The ManagerRef class depends on the IWorker abstraction, not specific
    %   worker implementations.
    % - This adheres to DIP, allowing the ManagerRef class to work with any class
    %   that implements IWorker.

    manager = ManagerRef();
    manager.addWorker(WorkerRef());
    manager.addWorker(RobotRef());
    manager.manage();
end

classdef IWorker < handle
    methods (Abstract)
        work(obj)
    end
end

classdef WorkerRef < IWorker
    methods
        function work(obj)
            fprintf('Worker is working...\n');
        end
    end
end

classdef RobotRef < IWorker
    methods
        function work(obj)
            fprintf('Robot is working...\n');
        end
    end
end

classdef ManagerRef < handle
    properties (Access = private)
        workers = {}
    end
    methods
        function addWorker(obj, worker)
            obj.workers{end+1} = worker;
        end
        function manage(obj)
            for i = 1:length(obj.workers)
                worker = obj.workers{i};
                worker.work();
            end
        end
    end
end
